src/table/use-sticky-columns.ts

```diff
@@ -0,0 +1,362 @@
+        const className = getClassName(state);
+        const cellElement = cellRef.current;
+        if (cellElement) {
+          Object.keys(className).forEach(key => {
+            if (className[key]) {
+              cellElement.classList.add(key);
+            } else {
+              cellElement.classList.remove(key);
+            }
+          });
+          cellElement.style.left = state?.offset?.left !== undefined ? `${state?.offset?.left}px` : '';
+          cellElement.style.right = state?.offset?.right !== undefined ? `${state?.offset?.right}px` : '';
+        }
+      };
+
+      const unsubscribe = stickyColumns.store.subscribe(selector, newState => updateCellStyles(selector(newState)));
+      return unsubscribe;
+    },
+    // getClassName is expected to be pure
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [stickyColumns.store, stickyColumns.isEnabled, columnId]
+  );
+
+  // Provide cell styles as props so that a re-render won't cause invalidation.
+  const cellStyles = stickyColumns.store.get().cellState[columnId];
+  return {
+    ref: refCallback,
+    className: cellStyles ? clsx(getClassName(cellStyles)) : undefined,
+    style: cellStyles?.offset ?? undefined,
+  };
+}
+
+interface UpdateCellStylesProps {
+  wrapper: HTMLElement;
+  table: HTMLElement;
+  cells: Record<ColumnId, HTMLElement>;
+  visibleColumns: readonly ColumnId[];
+  stickyColumnsFirst: number;
+  stickyColumnsLast: number;
+}
+
+export default class StickyColumnsStore extends AsyncStore<StickyColumnsState> {
+  private cellOffsets = new Map<ColumnId, { first: number; last: number }>();
+  private stickyWidthLeft = 0;
+  private stickyWidthRight = 0;
+  private isStuckToTheLeft = false;
+  private isStuckToTheRight = false;
+
+  constructor() {
+    super({ cellState: {}, scrollPadding: { left: 0, right: 0 } });
+  }
+
+  public updateCellStyles(props: UpdateCellStylesProps) {
+    const hasStickyColumns = props.stickyColumnsFirst + props.stickyColumnsLast > 0;
+    const hadStickyColumns = this.cellOffsets.size > 0;
+
+    if (hasStickyColumns || hadStickyColumns) {
+      this.updateScroll(props);
+      this.updateCellOffsets(props);
+      this.set(() => ({
+        cellState: this.generateCellStyles(props),
+        scrollPadding: { left: this.stickyWidthLeft, right: this.stickyWidthRight },
+      }));
+    }
+  }
+
+  private updateScroll(props: UpdateCellStylesProps) {
+    const wrapperScrollLeft = props.wrapper.scrollLeft;
+    const wrapperScrollWidth = props.wrapper.scrollWidth;
+    const wrapperClientWidth = props.wrapper.clientWidth;
+    const tablePaddingLeft = parseInt(getComputedStyle(props.table).paddingLeft) || 0;
+    const tablePaddingRight = parseInt(getComputedStyle(props.table).paddingRight) || 0;
+
+    this.isStuckToTheLeft = wrapperScrollLeft > tablePaddingLeft;
+    this.isStuckToTheRight = Math.ceil(wrapperScrollLeft) < wrapperScrollWidth - wrapperClientWidth - tablePaddingRight;
```

Comment:
Why do we use `Math.ceil` here? This could have a risk of producing an incorrect result e.g. when the page is zoomed in and then scrolled to a non-integer scroll position at the end of the table

Reply:
When a table is used in its `full-page` variant in our AppLayout component, it sometimes produces the weird result where it's always "stuck to the right".
This is because - in certain browsers -  the `wrapperScrollLeft` value was a non-integer, lower than the element's `clientWidth` (by, for example, 0.5 px).
While I understand the risk you pointed out, if it does happen, I see it as being an edge-case with very low impact: the user would need to scroll 1 more pixel to be able to see the "stuck" state of the column.